File: src/mailer.py
import smtplib
from email.message import EmailMessage

def send_email(job_titles, sender_email, receiver_email, email_password):
    print("Sending email...")
    msg = EmailMessage()
    msg['Subject'] = f'AutoWork: Found {len(job_titles)} Jobs'
    msg['From'] = f"autowork <{sender_email}>"
    msg['To'] = receiver_email

    # Plain text fallback
    text_body = f"NEW jobs found on Workday ({len(job_titles)} total):\n\n"
    
    # Rich HTML formatted version
    html_body = f"""
    <html>
      <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; color: #333; line-height: 1.6;">
        <h2 style="color: #004d99; border-bottom: 2px solid #f0f0f0; padding-bottom: 10px;"> AutoWork Jobs Report</h2>
        <p style="font-size: 15px;">We found <b>{len(job_titles)}</b> completely new job postings on Workday.</p>
        <div style="margin-top: 20px;">
    """
    
    for idx, job in enumerate(job_titles, 1):
        date_str = job.posting_date.strftime("%b %d, %Y") if job.posting_date else "Unknown Date"
        
        # Build plain text variant
        text_body += f"{job.name}\n   ID: {job.id} | Posted: {date_str}\n   Location: {job.location_city}, {job.location_country}\n\n"
        
        # Build HTML card variant
        html_body += f"""
          <div style="margin-bottom: 16px; padding: 18px; border: 1px solid #e1e4e8; border-radius: 8px; background-color: #fcfcfc;">
              <h3 style="margin: 0 0 8px 0; color: #111; font-size: 17px;">{job.name}</h3>
              <div style="font-size: 14px; color: #555;">
                  <span style="color: #777;">ID:</span> {job.id} &nbsp;|&nbsp; 
                  <span style="color: #777;">Posted:</span> {date_str}<br>
                  <span style="color: #777;">Location:</span> {job.location_city}, {job.location_country}
              </div>
          </div>
        """
        
    html_body += """
        </div>
      </body>
    </html>
    """
    
    # Package into the EmailMessage
    msg.set_content(text_body)
    msg.add_alternative(html_body, subtype='html')

    # Example uses Gmail's SMTP server. 
    # If using Outlook/Office365, use smtp.office365.com and port 587
    try:
        with smtplib.SMTP_SSL('smtp.gmail.com', 465) as smtp:
            smtp.login(sender_email, email_password)
            smtp.send_message(msg)
        print("Email sent successfully!")
    except Exception as e:
        print(f"Failed to send email: {e}")

def send_duo_email(code, sender_email, receiver_email, email_password):
    print("Emailing Duo verification code...")
    msg = EmailMessage()
    msg['Subject'] = f'URGENT: Duo Code {code}'
    msg['From'] = f"autowork <{sender_email}>"
    msg['To'] = receiver_email

    # Plain text fallback
    text_body = f"Your Workday automation requires Duo authentication.\n\nPlease open your Duo Mobile App and enter: {code}\n\nYou have less than 90 seconds before the login times out."
    
    # Rich HTML formatted version
    html_body = f"""
    <html>
      <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; color: #333; line-height: 1.6; padding: 20px; background-color: #f7f7f7;">
        <div style="max-width: 380px; margin: 0 auto; border: 1px solid #e1e4e8; border-radius: 12px; background-color: #ffffff; padding: 35px 25px; text-align: center; box-shadow: 0 4px 6px rgba(0,0,0,0.05);">
            <h2 style="color: #d93025; margin-top: 0; font-size: 22px;">🚨 Auth Required</h2>
            <p style="font-size: 15px; color: #555; margin-bottom: 25px;">Your AutoWork tracker requires Duo mobile verification to continue logging in.</p>
            
            <div style="margin: 0; padding: 20px; background-color: #f8f9fa; border-radius: 8px; border: 1px solid #e9ecef;">
                <p style="margin: 0; font-size: 13px; text-transform: uppercase; color: #6c757d; letter-spacing: 1.5px; font-weight: bold;">Verification Code</p>
                <h1 style="margin: 10px 0 0 0; font-size: 54px; letter-spacing: 6px; color: #212529;">{code}</h1>
            </div>
            
            <p style="font-size: 13px; color: #777; margin-top: 25px; margin-bottom: 0;">⏳ Please enter this in your Duo app immediately. The browser will time out.</p>
        </div>
      </body>
    </html>
    """
    
    msg.set_content(text_body)
    msg.add_alternative(html_body, subtype='html')

    try:
        with smtplib.SMTP_SSL('smtp.gmail.com', 465) as smtp:
            smtp.login(sender_email, email_password)
            smtp.send_message(msg)
        print("Duo email sent successfully!")
    except Exception as e:
        print(f"Failed to send Duo email: {e}")

def send_error_email(error_message, sender_email, receiver_email, email_password, screenshot_path=None):
    print("Emailing error notification...")
    msg = EmailMessage()
    msg['Subject'] = '⚠️ AutoWork: Error Encountered'
    msg['From'] = f"autowork <{sender_email}>"
    msg['To'] = receiver_email

    # Plain text fallback
    text_body = f"Your Workday automation encountered an error:\n\n{error_message}\n\nPlease check the server logs."
    
    # Rich HTML formatted version
    html_body = f"""
    <html>
      <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; color: #333; line-height: 1.6; padding: 20px; background-color: #f7f7f7;">
        <div style="max-width: 500px; margin: 0 auto; border: 1px solid #e1e4e8; border-radius: 12px; background-color: #ffffff; padding: 35px 25px; box-shadow: 0 4px 6px rgba(0,0,0,0.05);">
            <h2 style="color: #d93025; margin-top: 0; font-size: 22px;">⚠️ Automation Error</h2>
            <p style="font-size: 15px; color: #555; margin-bottom: 25px;">Your AutoWork tracker ran into an issue and crashed.</p>
            
            <div style="margin: 0; padding: 15px; background-color: #fff3f3; border-radius: 8px; border: 1px solid #ffcdd2;">
                <p style="margin: 0; font-size: 13px; text-transform: uppercase; color: #d32f2f; letter-spacing: 1.5px; font-weight: bold;">Error Traceback</p>
                <pre style="margin: 10px 0 0 0; font-size: 13px; color: #333; white-space: pre-wrap; word-wrap: break-word;">{error_message}</pre>
            </div>
            
            <p style="font-size: 13px; color: #777; margin-top: 25px; margin-bottom: 0;">Please VPN into your Ubuntu server and check the terminal logs.</p>
        </div>
      </body>
    </html>
    """
    
    msg.set_content(text_body)
    msg.add_alternative(html_body, subtype='html')

    if screenshot_path:
        try:
            with open(screenshot_path, 'rb') as f:
                img_data = f.read()
            msg.add_attachment(img_data, maintype='image', subtype='png', filename='error_screenshot.png')
        except Exception as e:
            print(f"Could not attach screenshot: {e}")

    try:
        with smtplib.SMTP_SSL('smtp.gmail.com', 465) as smtp:
            smtp.login(sender_email, email_password)
            smtp.send_message(msg)
        print("Error email sent successfully!")
    except Exception as e:
        print(f"Failed to send Error email: {e}")

File: src/mailer.rs
use lettre::message::{Attachment, MultiPart, SinglePart, header::ContentType};
use lettre::{Message, SmtpTransport, Transport};
use crate::models::Job;

pub fn send_email(job_titles: &[Job], sender_email: &str, receiver_emails: &[String], email_password: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("Sending email...");
    
    let mut text_body = format!("NEW jobs found on Workday ({} total):\n\n", job_titles.len());
    let mut html_body = format!(
        "<html><body style=\"font-family: sans-serif;\"><h2>AutoWork Jobs Report</h2><p>Found <b>{}</b> new jobs.</p><div>",
        job_titles.len()
    );

    for job in job_titles {
        let date_str = match &job.posting_date {
            Some(d) => d.format("%b %d, %Y").to_string(),
            None => "Unknown Date".to_string(),
        };

        let url_text = if job.url.is_empty() {
            "".to_string()
        } else {
            format!("\n   URL: {}", job.url)
        };
        
        let desc_text = if let Some(summary) = &job.ai_summary {
            let mut s = String::from("\n   AI Summary:\n");
            s.push_str(&format!("   - Availability: {}\n", summary.availability));
            s.push_str(&format!("   - Pay Rate: {}\n", summary.pay_rate));
            s.push_str(&format!("   - Responsibilities: {}\n", summary.responsibilities));
            s.push_str(&format!("   - Eligibility: {}\n", summary.eligibility));
            s.push_str(&format!("   - Note: {}\n", summary.note));
            s
        } else if let Some(ref d) = job.description {
            format!("\n   Description:\n   {}\n", d.replace("\n", "\n   "))
        } else {
            "".to_string()
        };

        text_body.push_str(&format!("{}\n   ID: {} | Posted: {}\n   Location: {}, {}{}{}\n\n", 
            job.name, job.id, date_str, job.location_city, job.location_country, url_text, desc_text));
            
        let url_html = if job.url.is_empty() {
            "".to_string()
        } else {
            format!("<br><a href=\"{}\" target=\"_blank\">View Job Posting</a>", job.url)
        };
        
        let desc_html = if let Some(summary) = &job.ai_summary {
            let mut s = String::from("<div style=\"margin-top:12px;color:#333;font-size:14px;background:#f9f9f9;padding:10px;border-radius:5px;\"><strong>✨ AI Summary:</strong><ul style=\"margin-top:8px;margin-bottom:0;padding-left:20px;\">");
            s.push_str(&format!("<li><strong>Availability</strong>: {}</li>", summary.availability));
            s.push_str(&format!("<li><strong>Pay Rate</strong>: {}</li>", summary.pay_rate));
            s.push_str(&format!("<li><strong>Responsibilities</strong>: {}</li>", summary.responsibilities));
            s.push_str(&format!("<li><strong>Eligibility</strong>: {}</li>", summary.eligibility));
            s.push_str(&format!("<li><strong>Note</strong>: {}</li>", summary.note));
            s.push_str("</ul></div>");
            s
        } else if let Some(ref d) = job.description {
            format!("<div style=\"margin-top:12px;color:#555;font-size:14px;\">{}</div>", d.replace("\n", "<br>"))
        } else {
            "".to_string()
        };

        html_body.push_str(&format!(
            "<div style=\"margin-bottom:16px;padding:18px;border:1px solid #e1e4e8;\"><h3>{}</h3><p>ID: {} | Posted: {}<br>Location: {}, {}{}</p>{}</div>",
            job.name, job.id, date_str, job.location_city, job.location_country, url_html, desc_html
        ));
    }
    html_body.push_str("</div></body></html>");

    let to_addresses = receiver_emails.join(", ");

    let email = Message::builder()
        .from(sender_email.parse()?)
        .to(to_addresses.parse()?)
        .subject(format!("AutoWork: Found {} Jobs", job_titles.len()))
        .multipart(
            MultiPart::alternative()
                .singlepart(SinglePart::plain(text_body))
                .singlepart(SinglePart::html(html_body))
        )?;

    let mailer = SmtpTransport::relay("smtp.gmail.com")?
        .credentials(lettre::transport::smtp::authentication::Credentials::new(
            sender_email.to_string(),
            email_password.to_string(),
        ))
        .build();

    mailer.send(&email)?;
    println!("Email sent successfully!");
    Ok(())
}

pub fn send_duo_email(code: &str, sender_email: &str, receiver_email: &str, email_password: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("Emailing Duo verification code...");
    
    let text_body = format!("Your Workday automation requires Duo authentication. Code: {}", code);
    let html_body = format!("<html><body><h2>🚨 Auth Required</h2><h1>{}</h1><p>Enter in your Duo app immediately.</p></body></html>", code);

    let email = Message::builder()
        .from(sender_email.parse()?)
        .to(receiver_email.parse()?)
        .subject(format!("URGENT: Duo Code {}", code))
        .multipart(
            MultiPart::alternative()
                .singlepart(SinglePart::plain(text_body))
                .singlepart(SinglePart::html(html_body))
        )?;

    let mailer = SmtpTransport::relay("smtp.gmail.com")?
        .credentials(lettre::transport::smtp::authentication::Credentials::new(
            sender_email.to_string(),
            email_password.to_string(),
        ))
        .build();

    mailer.send(&email)?;
    println!("Duo email sent successfully!");
    Ok(())
}

pub fn send_error_email(error_message: &str, screenshot: Option<Vec<u8>>, sender_email: &str, receiver_email: &str, email_password: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("Emailing error notification...");
    
    let text_body = format!("Error Encountered:\n\n{}", error_message);
    let html_body = format!("<html><body><h2>⚠️ Automation Error</h2><pre>{}</pre></body></html>", error_message);

    let multipart = MultiPart::alternative()
        .singlepart(SinglePart::plain(text_body))
        .singlepart(SinglePart::html(html_body));

    let email_builder = Message::builder()
        .from(sender_email.parse()?)
        .to(receiver_email.parse()?)
        .subject("⚠️ AutoWork: Error Encountered");

    let email = if let Some(png_data) = screenshot {
        let attachment = Attachment::new(String::from("error_screenshot.png"))
            .body(png_data, ContentType::parse("image/png").unwrap());
        email_builder.multipart(MultiPart::mixed().multipart(multipart).singlepart(attachment))?
    } else {
        email_builder.multipart(multipart)?
    };

    let mailer = SmtpTransport::relay("smtp.gmail.com")?
        .credentials(lettre::transport::smtp::authentication::Credentials::new(
            sender_email.to_string(),
            email_password.to_string(),
        ))
        .build();

    mailer.send(&email)?;
    println!("Error email sent successfully!");
    Ok(())
}

pub fn send_debug_screenshot_email(png_data: Vec<u8>, sender_email: &str, receiver_email: &str, email_password: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("Emailing debug screenshot...");
    
    let text_body = "Debug Screenshot Attached.".to_string();
    let html_body = "<html><body><h2>🐛 Debug Screenshot</h2><p>Here is the latest screenshot from the headless browser.</p></body></html>".to_string();

    let multipart = MultiPart::alternative()
        .singlepart(SinglePart::plain(text_body))
        .singlepart(SinglePart::html(html_body));

    let attachment = Attachment::new(String::from("debug_screenshot.png"))
        .body(png_data, ContentType::parse("image/png").unwrap());

    let email = Message::builder()
        .from(sender_email.parse()?)
        .to(receiver_email.parse()?)
        .subject("🐛 AutoWork Debug Screenshot")
        .multipart(MultiPart::mixed().multipart(multipart).singlepart(attachment))?;

    let mailer = SmtpTransport::relay("smtp.gmail.com")?
        .credentials(lettre::transport::smtp::authentication::Credentials::new(
            sender_email.to_string(),
            email_password.to_string(),
        ))
        .build();

    mailer.send(&email)?;
    println!("Debug screenshot emailed successfully!");
    Ok(())
}
